# ENGRI 1120: Series Cell-free Production and Recovery of the mRNA BNT-162b2 Vaccine

Rakshith Sreeram, rs2244, Xander Marroquin, acm336, Gavin McCarthy, gm548

<img src="figs/ChemE project-1" style="width:50%">

## Introduction
Since the initial outbreak in Wuhan China on January 10th, 2020, governments around the world have been attempting to pioneer vaccines to combat the Covid outbreak. What seems to be the solution, mRNA vaccines work by introducing the body to mRNA rather than dead parts of bacteria or viruses. This in turn induces the immune system to produce antibodies against the corresponding virus proteins that the mRNA is a replica of. While mRNA has been used to combat the coronavirus, the virus has mutated many times and requires a robust system to manufacture pure mRNA for the vaccine in sufficient quantities to vaccinate the general population. Additional concerns regarding the production of the mRNA vaccine include the low-temperature storage that is required to stabilize mRNA. Additionally, a gene must be introduced along with nucleotides in order to facilitate the creation of mRNA.

The goal was to efficiently produce mRNA by the reaction between Gene and Pure express which is concentrated with 100 μM of T7RNAP, M_atp_c, M_utp_c, M_ctp_c, M_gtp_c. The process starts off with two streams, stream one contains gene and the other containing Pure-Express. A syringe pump is installed at the beginning of the streams in order to feed the reactors at a sufficient flow rate. A splitter on stream 1 splits the stream into 16 different streams, each going into a separate reactor chip in order to allow for a series configuration of reactor chips. This configuration allows us to maximize the reaction extent in order to reach the goal of producing a concentration of 100 micromolar of mRNA. After the reactions occur via our series configuration of 16 reactor chips, the mRNA is purified to a 99.994% via a setup of 5 99% separators and cooled from 25 degrees celsius to -25 degrees celsius for storage through the use of a heat exchanger so the mRNA can be stabilized.

## Materials and Methods

### Project Setup

In [1]:
import Pkg; Pkg.activate("."); Pkg.resolve(); Pkg.instantiate();

  Activating project at `C:\Users\xande\Downloads\ENGRI-1120-IntroToChemE-Example-Notebooks-main (7)\ENGRI-1120-IntroToChemE-Example-Notebooks-main\project`
    Updating `C:\Users\xande\Downloads\ENGRI-1120-IntroToChemE-Example-Notebooks-main (7)\ENGRI-1120-IntroToChemE-Example-Notebooks-main\project\Project.toml`
  [5ae59095] + Colors v0.12.8
  [a93c6f00] + DataFrames v1.3.4
  [5789e2e9] + FileIO v1.15.0
  [60bf3e95] + GLPK v1.1.0
  [033835bb] + JLD2 v0.4.22
  [91a5bcdd] + Plots v1.32.0
  [08abe8d2] + PrettyTables v1.3.1
    Updating `C:\Users\xande\Downloads\ENGRI-1120-IntroToChemE-Example-Notebooks-main (7)\ENGRI-1120-IntroToChemE-Example-Notebooks-main\project\Manifest.toml`
  [79e6a3ab] + Adapt v3.4.0
  [6e4b80f9] + BenchmarkTools v1.3.1
  [d360d2e6] + ChainRulesCore v1.15.4
  [9e997f8a] + ChangesOfVariables v0.1.4
  [523fee87] + CodecBzip2 v0.7.2
  [944b1d66] + CodecZlib v0.7.0
  [35d6a980] + ColorSchemes v3.19.0
  [3da002f7] + ColorTypes v0.11.4
  [c3611d14] + ColorVectorSpace v

  [29816b5a] + LibSSH2_jll v1.10.2+0
  [c8ffd9c3] + MbedTLS_jll v2.28.0+0
  [14a3606d] + MozillaCACerts_jll v2022.2.1
  [4536629a] + OpenBLAS_jll v0.3.20+0
  [05823500] + OpenLibm_jll v0.8.1+0
  [83775a58] + Zlib_jll v1.2.12+3
  [8e850b90] + libblastrampoline_jll v5.1.1+0
  [8e850ede] + nghttp2_jll v1.48.0+0
  [3f19e933] + p7zip_jll v17.4.0+0
        Info Packages marked with ⌅ have new versions available but cannot be upgraded. To see why use `status --outdated -m`


In [2]:
# load reqd packages and set paths -
using JLD2
using FileIO
using PrettyTables
using DataFrames
using GLPK

# setup paths -
const _ROOT = pwd();
const _PATH_TO_DATA = joinpath(_ROOT, "data");

#### Load the project code library
The call to the `include` function loads the `ENGRI-1120-Project-CodeLib.jl` library into the notebook; the library contains functions we can use during the project. In particular, it includes the function:

* The `compute_optimal_extent(stoichiometric_matrix::Array{Float64,2}, default_bounds_array::Array{Float64,2},
    species_bounds_array::Array{Float64,2}, objective_coefficient_array::Array{Float64,1}; min_flag::Bool = true) -> Tuple` function calls the [GLPK](https://www.gnu.org/software/glpk/) linear program solver. The `results` tuple contains several things, but the important ones are `calculated_flux_array`, `objective_value`, and the status/exit flags `status_flag` and `exit_flag` (which let us know if the solver successfully found a solution).
* The `build(model::Type{MSULatticeModel}; ṅₒ::Float64, L::Int64, u::Float64, d::Float64) -> MSULatticeModel` function builds a [Binary tree](https://en.wikipedia.org/wiki/Binary_tree) of Magical Separation Units (MSUs). Arguments: $ṅₒ$ denotes the species mole flow rate into the separation system, $L$ denotes the number of layers of the tree, $u$ denotes the `up` factor (split for the `up` path), and $d$ denotes the `down` factor (the split for the `down` path). This function returns the `MSULatticeModel` model, which contains the column array `data` holding the species mole flow rate for each node in the tree. 
* The `build_nodes_dictionary(levels::Int64) -> Dict{Int64,Array{Int64,1}}` function constructs a dictionary of node indexes for each level of the tree; keys are the tree levels.

In [3]:
include("ENGRI-1120-Project-CodeLib.jl");

In [4]:
# load the model file -
model = load(joinpath(_PATH_TO_DATA, "ENGRI-1120-BNT162b2-Model.jld2"))["model"]

┌ Warning: the following fields are present in type DataFrame saved in the file but not present in the type the workspace:
│ 
│ metadata
│ colmetadata
│ allnotemetadata,
│ 
│ Data in these fields will not be accessible
└ @ JLD2 C:\Users\xande\.julia\packages\JLD2\k9Gt0\src\data\reconstructing_datatypes.jl:196


Dict{String, Any} with 7 entries:
  "stochiometric_matrix" => [-1.0 0.0 … 0.0 -1.0; -1.0 0.0 … -1.0 0.0; … ; 1.0 …
  "list_of_reactions"    => ["TX_BNT_162b2_binding", "TX_BNT_162b2_open", "BNT_…
  "reaction_table"       => 6×7 DataFrame…
  "flux_bounds_array"    => [-1000.0 1000.0; 0.0 1000.0; … ; 0.0 1000.0; 0.0 10…
  "mRNA_sequence"        => ['C', 'U', 'C', 'U', 'U', 'A', 'U', 'U', 'U', 'G'  …
  "list_of_species"      => ["G_BNT_162b2", "T7RNAP", "M_atp_c", "M_utp_c", "M_…
  "gene_sequence"        => ['G', 'A', 'G', 'A', 'A', 'T', 'A', 'A', 'A', 'C'  …

In [5]:
# get stuff from the model data structure -
S = model["stochiometric_matrix"]; # fix the spelling in the model file
flux_bounds_array = model["flux_bounds_array"];
list_of_species = model["list_of_species"];
list_of_reactions = model["list_of_reactions"];
reaction_table = model["reaction_table"];
gene_sequence = model["gene_sequence"];

In [6]:
reaction_table

6×7 DataFrame
 Row │ id                          forward                            reverse  ⋯
     │ String                      String                             String   ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ TX_BNT_162b2_binding        G_BNT_162b2+T7RNAP                 G_BNT_16 ⋯
   2 │ TX_BNT_162b2_open           G_BNT_162b2_T7RNAP_closed          G_BNT_16
   3 │ BNT_162b2_transcription     G_BNT_162b2_T7RNAP_open+798*M_at…  mRNA_BNT
   4 │ mRNA_BNT_162b2_degradation  mRNA_BNT_162b2                     798*M_am
   5 │ RNAP_deactivation           T7RNAP                             T7RNAP_i ⋯
   6 │ GENE_deactivation           G_BNT_162b2                        G_BNT_16
                                                               5 columns omitted

In [7]:
# How many species and reactions?
(ℳ, ℛ) = size(S);

In [8]:
# initialize -
species_index_table_data = Array{Any,2}(undef, ℳ, 2);

# build table -
for i ∈ 1:ℳ
    species_index_table_data[i,1] = i;
    species_index_table_data[i,2] = list_of_species[i];
end

# setup header -
species_index_header_table = (["Index", "Species"]);

# build table -
pretty_table(species_index_table_data; header=species_index_header_table);

┌───────┬───────────────────────────┐
│ Index │                   Species │
├───────┼───────────────────────────┤
│     1 │               G_BNT_162b2 │
│     2 │                    T7RNAP │
│     3 │                   M_atp_c │
│     4 │                   M_utp_c │
│     5 │                   M_ctp_c │
│     6 │                   M_gtp_c │
│     7 │            mRNA_BNT_162b2 │
│     8 │                   M_ppi_c │
│     9 │                   M_amp_c │
│    10 │                   M_ump_c │
│    11 │                   M_cmp_c │
│    12 │                   M_gmp_c │
│    13 │           T7RNAP_inactive │
│    14 │      G_BNT_162b2_inactive │
│    15 │ G_BNT_162b2_T7RNAP_closed │
│    16 │   G_BNT_162b2_T7RNAP_open │
└───────┴───────────────────────────┘


#### Setup the constants, feed rates and compositions

In [48]:
# how many chips in series?
number_of_chips = 16;

# what fraction of the mRNA degrades?
δ = 0.10;

# MSU split ratio -
θ = 0.99;

# Setup constants for transcription -
L = length(gene_sequence);
K = 0.116; # saturation constant units: μmol/L; Source: ACS Synth. Biol. 2018, 7, 8, 1844–1857 https://doi.org/10.1021/acssynbio.7b00465
v̇ₜ = (90.0)*(60); # units: nt/s; Source: BIND: 111871
u = 0.95; # u-factor; Source: ACS Synth. Biol. 2018, 7, 8, 1844–1857 https://doi.org/10.1021/acssynbio.7b00465

# volume -
V = 100.0*(1/1e6); # liquid reaction volume on each chip units: L

# Stock solution: PURExpress -> fed into chips by pump 2 (stream 2)
# PURExpress flows into the chip in stream 2
T7RNAP = 100.0;          # concentration in PURExpress units: μmol/L
M_atp_c = 100*(1e6/1e3); # concentration in PURExpress units: μmol/L
M_utp_c = 100*(1e6/1e3); # concentration in PURExpress units: μmol/L
M_ctp_c = 100*(1e6/1e3); # concentration in PURExpress units: μmol/L
M_gtp_c = 100*(1e6/1e3); # concentration in PURExpress units: μmol/L

# Stock solution: DNA -> feed into splitter by pump 1 and then into chip (always stream 1)
G_BNT_162b2 = 1.0;     # gene concentration in stock solution units: μmol/L

# Volumetric flow rates from the pump *into* the splitter unit or chip - our base case will 1 ml/min; thus, let's
# scale by the number of chips
Ḟ₁ = number_of_chips*1000.0*(1/1e6); # volumetric flow rate of syringe pump 1 units: L/min
Ḟ₂ = number_of_chips*1000.0*(1/1e6); # volumetric flow rate of syringe pump 2 units: L/min

# stuff for the tables -
current_table_counter = 0; # do not change me

#### Specify the inputs streams

##### a) Specify the composition of feed stream 1
By default, this stream contains only the gene encoding the mRNA BNT-162b2 product. The mainstream is split into sub-streams that are fed into each chip.

In [49]:
# setup feed compostions for feed stream 1 
ṅ₁ = zeros(ℳ); # default is zero, correct specific values -
ṅ₁[1] = G_BNT_162b2*Ḟ₁*(1/number_of_chips); # units: μmol/min

##### b) Specify the composition of feed stream 2
By default, feed stream 2 contains the [PURExpress](https://www.neb.com/products/e6800-purexpress-invitro-protein-synthesis-kit#Product%20Information), which has everything we need to make our mRNA product of interest _expect_ the linear DNA. 

In [50]:
# setup feed compostions for feed stream 2: this feed goes into chip 1
ṅ₂ = zeros(ℳ); # default is zero, then correct specific values -
ṅ₂[2] = T7RNAP*Ḟ₂;  # units: μmol/min
ṅ₂[3] = M_atp_c*Ḟ₂; # units: μmol/min
ṅ₂[4] = M_utp_c*Ḟ₂; # units: μmol/min
ṅ₂[5] = M_ctp_c*Ḟ₂; # units: μmol/min
ṅ₂[6] = M_gtp_c*Ḟ₂; # units: μmol/min

### Flux balance analysis setup
* Describe the flux balance analysis method. 
* How did your group update the default species and flux bounds to model ths process?
* What objective function did you choose?

#### Species bounds

In [51]:
# species bounds array -
species_bounds_array = [-(ṅ₁ .+ ṅ₂) 10000.0*ones(ℳ,1)];

#### Reaction bounds
In the series case, reaction bounds will change for every chip. Let's set a default upper bound and then update the other bounds below. If you are a brave soul, you can alter the per-chip bounds, but you don't need to do so.

In [52]:
# initialize -
flux_bounds_array = zeros(ℛ,2);
flux_bounds_array[:,2] .= 1000.0; # large default upper bound

In [53]:
# initialize -
reaction_index_table_data = Array{Any,2}(undef, ℛ, 4);

# build table -
for i ∈ 1:ℛ
    reaction_index_table_data[i,1] = i;
    reaction_index_table_data[i,2] = list_of_reactions[i];
    reaction_index_table_data[i,3] = flux_bounds_array[i,1];
    reaction_index_table_data[i,4] = flux_bounds_array[i,2];
end

# setup title string -
reaction_table_title = "Table $(current_table_counter+=1): Reaction index-name mapping table."

# setup header -
reaction_index_header_table = (["Index", "Reaction", "lower bound", "upper bound"], ["", "", "μmol/min", "μmol/min"]);

# build table -
pretty_table(reaction_index_table_data, title=reaction_table_title; header=reaction_index_header_table);

Table 1: Reaction index-name mapping table.
┌───────┬────────────────────────────┬─────────────┬─────────────┐
│ Index │                   Reaction │ lower bound │ upper bound │
│       │                            │    μmol/min │    μmol/min │
├───────┼────────────────────────────┼─────────────┼─────────────┤
│     1 │       TX_BNT_162b2_binding │         0.0 │      1000.0 │
│     2 │          TX_BNT_162b2_open │         0.0 │      1000.0 │
│     3 │    BNT_162b2_transcription │         0.0 │      1000.0 │
│     4 │ mRNA_BNT_162b2_degradation │         0.0 │      1000.0 │
│     5 │          RNAP_deactivation │         0.0 │      1000.0 │
│     6 │          GENE_deactivation │         0.0 │      1000.0 │
└───────┴────────────────────────────┴─────────────┴─────────────┘


#### Objective coefficient array
* What is the objective you used?

In [54]:
# setup the objective coefficient array -
obj_vector = zeros(ℛ);
obj_vector[3] = -1; # why negative?

## Results and Discussion 

### Compute the optimal extent of reaction and exit stream for first chip

In [55]:
# initialize some space to store tmp soln -
tmp_sim_storage_space = Dict{Int64,Tuple{Array{Float64,1},Array{Float64,1}}}();

# compute -
ṅ₃ = zeros(ℳ);
for chip_index ∈ 1:number_of_chips
    
    if (chip_index == 1)
        ṅ_chip = ṅ₂
    else
        ṅ_chip = ṅ₃
    end
        
    # update the bounds model -
    F̂₁ = ((chip_index/number_of_chips)*Ḟ₁)/(Ḟ₂+(chip_index/number_of_chips)*Ḟ₁)
    F̂₂ = Ḟ₂/(Ḟ₂+(chip_index/number_of_chips)*Ḟ₁);
    
    # Estimate the RNAP and GENE concentration -
    Rₜ = T7RNAP*F̂₂;                          # effective RNAP concentratation on the chip for the bounds units: μmol/L
    GENE = G_BNT_162b2*F̂₁ + G_BNT_162b2*F̂₂  # effective GENE concentratation on the chip for the bounds units: μmol/L
    flux_bounds_array[3,:] .= Rₜ*(v̇ₜ/L)*u*(GENE/(K+GENE))*V; # equality constraint

    # Setup bound for degradation (lower bound)
    flux_bounds_array[4,1] = δ*flux_bounds_array[3,1];
    
    # compute the optimal flux, and then estimate the output on the chip
    result = compute_optimal_extent(S, flux_bounds_array, species_bounds_array, obj_vector);
    ϵ̇ = result.calculated_flux_array;
    ṅ₃ = (ṅ₁ + ṅ_chip) + S*ϵ̇;   # compute the output from the chip -
    
    # now, the outlet from chip i-1 is the input to chip 1 (in stream 2), but stream 1 stays the same
    species_bounds_array = [-(ṅ₁ .+ ṅ_chip) 10000.0*ones(ℳ,1)];
    
    # grab -
    tmp_sim_storage_space[chip_index] = (ṅ₃,ϵ̇);
end

In [56]:
system_flow_table_data = Array{Any,2}(undef, ℳ, number_of_chips+2);

# populate the table -
for i ∈ 1:ℳ
    system_flow_table_data[i,1] = list_of_species[i];
    system_flow_table_data[i,2] = i;
    
    for chip_index ∈ 1:number_of_chips
        ṅ₃ = tmp_sim_storage_space[chip_index][1];
        system_flow_table_data[i,chip_index+2] = round(ṅ₃[i], digits=3);
    end
end

# build header -
# names 
flow_table_name_row = Array{String,1}()
push!(flow_table_name_row, "Species");
push!(flow_table_name_row, "index i")
for chip_index ∈ 1:number_of_chips
    push!(flow_table_name_row, "Chip $(chip_index)")
end

# units
flow_table_units_row = Array{String,1}()
push!(flow_table_units_row, "");
push!(flow_table_units_row, "")
for chip_index ∈ 1:number_of_chips
    push!(flow_table_units_row, "μmol/min")
end


# title -
flow_table_title = "Table $(current_table_counter+=1): Chip species mole flow rates table"

# header -
flux_table_header = (flow_table_name_row, flow_table_units_row)

# show -
pretty_table(system_flow_table_data, title=flow_table_title; header=flux_table_header)

Table 2: Chip species mole flow rates table
┌───────────────────────────┬─────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┐
│                   Species │ index i │   Chip 1 │   Chip 2 │   Chip 3 │   Chip 4 │   Chip 5 │   Chip 6 │   Chip 7 │   Chip 8 │   Chip 9 │  Chip 10 │  Chip 11 │  Chip 12 │  Chip 13 │  Chip 14 │  Chip 15 │  Chip 16 │
│                           │         │ μmol/min │ μmol/min │ μmol/min │ μmol/min │ μmol/min │ μmol/min │ μmol/min │ μmol/min │ μmol/min │ μmol/min │ μmol/min │ μmol/min │ μmol/min │ μmol/min │ μmol/min │ μmol/min │
├───────────────────────────┼─────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┤
│               G_BNT_162b2 │       1 │    0.001 │    0.002 │    0.003 │    0.004 │    0.005

In [57]:
# this will print the last few chips (in case the table formatting gets nutty)
system_flow_table_data

16×18 Matrix{Any}:
 "G_BNT_162b2"                 1     0.001  …     0.014     0.015     0.016
 "T7RNAP"                      2     1.6          1.6       1.6       1.6
 "M_atp_c"                     3  1591.73      1513.65   1509.11   1504.72
 "M_utp_c"                     4  1589.6       1491.36   1485.65   1480.12
 "M_ctp_c"                     5  1589.02      1485.3    1479.27   1473.44
 "M_gtp_c"                     6  1586.4    …  1458.03   1450.57   1443.35
 "mRNA_BNT_162b2"              7     0.009        0.097     0.103     0.107
 "M_ppi_c"                     8    43.253      451.669   475.388   498.367
 "M_amp_c"                     9     0.827        8.635     9.089     9.528
 "M_ump_c"                    10     1.04        10.864    11.435    11.988
 "M_cmp_c"                    11     1.098  …    11.47     12.073    12.656
 "M_gmp_c"                    12     1.36        14.197    14.943    15.665
 "T7RNAP_inactive"            13     0.0          0.0       0.0       0.0
 

### Design a downstream seperation using Magical Seperator Units (MSU)

In [58]:
# default -
ṅ₃ = tmp_sim_storage_space[number_of_chips][1]

# build a downstream seperation process with this number of levels:
number_of_levels = 5; # includes zero

In [59]:
# initialize -
tmp_storage_dict = Dict{Int64, MSULatticeModel}();

# is_product_vector -
is_product_vector = zeros(ℳ);
is_product_vector[7] = 1;

# compute the composition array -
for i ∈ 1:ℳ

    if (is_product_vector[i] == 1)
        msu_lattice_model = build(MSULatticeModel; ṅₒ = ṅ₃[i], L = number_of_levels , u = θ, d = (1 - θ));
    else
        msu_lattice_model = build(MSULatticeModel; ṅₒ = ṅ₃[i], L = number_of_levels , u = (1 - θ), d = θ);
    end
    
    # grab -
    tmp_storage_dict[i] = msu_lattice_model;
end

# grab the leaves -
nodes_dict = build_nodes_dictionary(number_of_levels);
children_dict = build_children_dictionary(nodes_dict);
tree_leaves = nodes_dict[number_of_levels-1];

# build a composition array -
number_of_nodes = length(tree_leaves);
composition_array = Array{Float64,2}(undef, number_of_nodes, ℳ);
for i ∈ 1:ℳ
    data = tmp_storage_dict[i].data;
    for j ∈ 1:number_of_nodes
        composition_array[j,i] = data[tree_leaves[j]]
    end
end

# make a pretty table and show the leaves of the tree -

# initialize -
sep_tree_flow_table_data = Array{Any,2}(undef, ℳ, length(tree_leaves) + 3)
for i ∈ 1:ℳ
    sep_tree_flow_table_data[i,1] = list_of_species[i];
    sep_tree_flow_table_data[i,2] = i;
    sep_tree_flow_table_data[i,3] = ṅ₃[i] # put node 0 in table -
        
    for j ∈ 1:length(tree_leaves)
        sep_tree_flow_table_data[i,3+j] = composition_array[j,i]
    end
end

# labels row -
label_row = Array{String,1}();
push!(label_row,"Species");
push!(label_row,"index i")
push!(label_row,"N0")
for j ∈ 1:length(tree_leaves)
    push!(label_row, "N$(tree_leaves[j])");
end

# units row -
units_row = Array{String,1}();
push!(units_row, "");
push!(units_row, "");
for j ∈ 1:length(tree_leaves)+1
    push!(units_row, "μmol/min");
end

# header -
sep_tree_flow_table_header = (label_row, units_row);

# set title -
title = "Table $(current_table_counter+=1): Magical Seperator Unit (MSU) flow table; N0 denotes the feed while N⋆ denotes the leaves of the tree."

# show -
pretty_table(sep_tree_flow_table_data, title=title; header=sep_tree_flow_table_header)

Table 3: Magical Seperator Unit (MSU) flow table; N0 denotes the feed while N⋆ denotes the leaves of the tree.
┌───────────────────────────┬─────────┬──────────┬────────────┬─────────────┬─────────────┬─────────────┬────────────┐
│                   Species │ index i │       N0 │        N11 │         N12 │         N13 │         N14 │        N15 │
│                           │         │ μmol/min │   μmol/min │    μmol/min │    μmol/min │    μmol/min │   μmol/min │
├───────────────────────────┼─────────┼──────────┼────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│               G_BNT_162b2 │       1 │    0.016 │    1.6e-10 │    1.584e-8 │  1.56816e-6 │ 0.000155248 │  0.0153695 │
│                    T7RNAP │       2 │      1.6 │     1.6e-8 │    1.584e-6 │ 0.000156816 │   0.0155248 │    1.53695 │
│                   M_atp_c │       3 │  1504.72 │ 1.50472e-5 │  0.00148967 │    0.147478 │     14.6003 │    1445.43 │
│                   M_utp_c │       4 │  1480.12 │ 1.480

In [60]:
# build mol frac composition table -

# construct mol frac array -
mol_frac_array = Array{Float64,2}(undef, ℳ, length(tree_leaves)+1);

# node 0 -
ṅ₃_total = sum(ṅ₃);
for i ∈ 1:ℳ
    mol_frac_array[i,1] = ṅ₃[i]*(1/ṅ₃_total);    
end

# get the sums along rows -
ṅ_total = sum(composition_array,dims = 2);
for node ∈ 1:length(tree_leaves)
    for i ∈ 1:ℳ
        mol_frac_array[i,node+1] = composition_array[node,i]*(1/ṅ_total[node]);
    end
end

# initialize -
sep_tree_mol_frac_table_data = Array{Any,2}(undef, ℳ, length(tree_leaves) + 3)
for i ∈ 1:ℳ
    sep_tree_mol_frac_table_data[i,1] = list_of_species[i];
    sep_tree_mol_frac_table_data[i,2] = i;
    sep_tree_mol_frac_table_data[i,3] = round(mol_frac_array[i,1], digits=4) # put node 0 in table -
        
    for j ∈ 1:length(tree_leaves)
        sep_tree_mol_frac_table_data[i, 3+j] = round(mol_frac_array[i,j+1], digits=4)
    end
end

# labels row -
label_mft_row = Array{String,1}();
push!(label_mft_row,"Species");
push!(label_mft_row,"index i")
push!(label_mft_row,"N0")
for j ∈ 1:length(tree_leaves)
    push!(label_mft_row, "N$(tree_leaves[j])");
end

# units row -
units_mft_row = Array{String,1}();
push!(units_mft_row, "");
push!(units_mft_row, "");
for j ∈ 1:length(tree_leaves)+1
    push!(units_mft_row, "mole frac");
end

# header -
sep_tree_mft_table_header = (label_mft_row, units_mft_row);

# set title -
title_mft = "Table $(current_table_counter+=1): Magical Seperator Unit (MSU) composition table."

# show -
pretty_table(sep_tree_mol_frac_table_data, title=title_mft; header=sep_tree_mft_table_header)

Table 4: Magical Seperator Unit (MSU) composition table.
┌───────────────────────────┬─────────┬───────────┬───────────┬───────────┬───────────┬───────────┬───────────┐
│                   Species │ index i │        N0 │       N11 │       N12 │       N13 │       N14 │       N15 │
│                           │         │ mole frac │ mole frac │ mole frac │ mole frac │ mole frac │ mole frac │
├───────────────────────────┼─────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────┤
│               G_BNT_162b2 │       1 │       0.0 │       0.0 │       0.0 │       0.0 │       0.0 │       0.0 │
│                    T7RNAP │       2 │    0.0002 │       0.0 │    0.0002 │    0.0002 │    0.0002 │    0.0002 │
│                   M_atp_c │       3 │    0.2332 │    0.0001 │    0.2005 │    0.2332 │    0.2332 │    0.2332 │
│                   M_utp_c │       4 │    0.2294 │    0.0001 │    0.1972 │    0.2294 │    0.2294 │    0.2294 │
│                   M_ctp_c │       5 │    0.22

### Financial analysis

Using the given financial information, we were able to calculate the cost of each component of the PURExpress using the given stock concentration and calculated flow rate. Having used 16 reactor chips, the syringe pump flow rate was calculated to be 0.016 L per min. Since the gene was synthesized in the lab, the projected cost was equal to 0. The PURExpress solution volume is equal to the flow rate (16000 μL per min). The rest of the components were calculated by multiplying the flow rate by the stock concentrations. Overall the total cost of the raw materials was 45,120 dollars per min with a flat fee of 5,175 dollars construction cost. The Net Present Value was calculated for 10 years using the total cost for 50 days of production. The NPV value came out to be -24,502,299,144.87 dollars (24.5 dollars B) which is less than 0 meaning the project will produce less income. In order to get the NPV to 0, we must set the price to be 3,248,640,686 dollars per year (3.2 dollars B annually), this means that the project will generate the same income as the alternative investment.

## Summary and conclusions
From our financial analysis, the annual operational costs of the production of mRNA significantly outweighs the startup cost associated with the respective equipment. As a result, the annual production cost can be brought down through the implementation of the 100% efficient separator. This brings up the startup cost because the current 5 99% separator system 2500 dollars while the 100% separator is 25000 dollars. However, this decreases the long-term annual production costs since none of the mRNA product is lost through subsequent separations. Another change that can be made throughout the process of producing mRNA would be to pump more Pure Express through a series of a two-three chips through a separate stream. This change would require the system to contain fewer reactor chips thus decreasing the start-up price but would significantly increase the annual cost as more Pure Express would need to be concentrated and run through the system.

## Video Summary and Description
https://drive.google.com/file/d/1YNvOeAk2voXkRjZ9cVu15B4Eh6S2_o2S/view?usp=drivesdk